Restructure production Docker scripts in #450.

rht opened the pull request:
This PR contains the entirety of #450, freshly restructured. There is a remaining catch-22 problem (fixable actually, see https://github.com/rht/zulip/commit/0b6ad2f28802770bdc520105470c651600a36646) that the production Dockerfile requires the tarball to contain the Puppet files in the first commit of this PR.

(cc: @galexrt)

smarx commented:
*Automated message from Dropbox CLA bot*

@rht, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
Can you extract out a commit at the bottom that moves the existing Dockerfile to Dockerfile.dev and updates the docs?  Seems like that we could just merge now.

Looking at the second-to-bottom commit, I'm a bit confused by `zulip_postsetup.conf`; why isn't that just a script if it's to be only run once?

timabbott commented:
@rht can you pull out the dockerfile dev stuff into their own PR?  I'd like Travis to pass for the pieces we're merging.

rht commented:
Passed in Travis for the dockerfile dev commits. These are at the bottom when seen from git instead of from here.

> Looking at the second-to-bottom commit, I'm a bit confused by zulip_postsetup.conf; why isn't that just a script if it's to be only run once?

It possibly made sense a year ago, because the Zulip server had to be running first before an admin user can be created. Tho from the log, `manage.py create_user` had existed a year ago.

Anyway, regardless of the hysteresis, I had added an amendment to the commits so that the zulip_postsetup.conf is now moved to a script that is run once at setup.

rht commented:
@timabbott if you merge the first 2 commits so that they are in 1.7, the subsequent commits will become simpler.

timabbott commented on puppet/zulip/files/supervisor/conf.d/zulip_postsetup.conf:
```diff
@@ -0,0 +1,9 @@
+stdout_events_enabled=true
+stderr_events_enabled=true
+autorestart = false
+exitcodes = 0,1,127
```
I don't understand why this command, which presumably should only be run once, is a supervisord config file.

rht replied:
(summoning @galexrt to provide the rationale)

rht replied:
I have just removed the run-only-once postsetup and rebased. @timabbott anything else needs to be fixed / rationalized?

timabbott commented on scripts/lib/create-zulip-admin:
```diff
@@ -0,0 +1,12 @@
+    exit 1
+fi
+set +e
+# Doing everything in python, even I never coded in python #YOLO
```
Let's remove this :)

timabbott commented on scripts/lib/create-zulip-admin:
```diff
@@ -0,0 +1,12 @@
+sudo -H -u zulip -g zulip bash <<BASH
+/home/zulip/deployments/current/manage.py create_user --this-user-has-accepted-the-tos --realm "$ZULIP_USER_DOMAIN" --password "$ZULIP_USER_PASS" "$ZULIP_USER_EMAIL" "$ZULIP_USER_FULLNAME"
+/home/zulip/deployments/current/manage.py knight "$ZULIP_USER_EMAIL" -f
+BASH
```
This script seems good.  Can you pull it into its own commit?

rht commented on scripts/lib/docker-functions.sh:
```diff
@@ -0,0 +1,409 @@
+    fi
+
+    if ([ "$ZULIP_USER_CREATION_ENABLED" = "True" ] || [ "$ZULIP_USER_CREATION_ENABLED" = "true" ]); then
+        /home/zulip/deployments/current/scripts/create-zulip-admin
```
I think the supervisord thing containing the create-zulip-admin can be removed given that it is already run here *once*.

rht commented on puppet/zulip/manifests/dockervoyager.pp:
```diff
@@ -0,0 +1,31 @@
+    mode    => "0644",
+    source  => "puppet:///modules/zulip/supervisor/conf.d/cron.conf",
+  }
+  file { "/etc/supervisor/conf.d/nginx.conf":
```
I think this can be used outside the docker scenario (e.g. when running puppet in vagrant).

rht commented:
(Is about to extract the first 2 commits to another PR for less hectic review.)

rht commented:
Down to 2 commits left !

timabbott commented:
@rht can you update this?  Now that we have certbot merged, I imagine this is getting close to being mergable.

rht commented:
I rebased. Currently, this uses self-signed certificate, which already makes it usable. I assume that the only remaining step is to change the cert to certbot-generated one?

rht commented:
@timabbott what is next at here? @gnprice?

timabbott commented:
@rht can you fix the tests and rebase issues, and give a report on how you've tested this?

timabbott commented:
Closing, since I think most of the work here is instead done in `zulip/docker-zulip`.
